### PR TITLE
Ethers V6: Add `SmartAccount` in order to provide better support for AA.

### PR DESCRIPTION
**Pull request by danijelTxFusion**
# What :computer: 
* Add `SmartAccount` in order to provide better AA support.

# Why :hand:

Based on #41, I continued to extend `SmartAccount` in order to improve  the support for account abstraction and paymaster.

`SmartAccount` is a class intended to provide better support for working with Account Abstraction and represents a kind of customized signer.

Instead of the `Signer` interface, `AbstractSigner` is used, which already has implementations of common base methods: (`estimateGas`, `getNonce`, `call`, etc). `SmartAccount` is implemented using the Template Design Pattern. `SmartAccount` contains a general flow for signing and sending transactions, with the possibility to customize certain parts of the functions. These functions are: 
- `transactionBuilder`: provides a way of writing arbitrary logic for populating zkSync EIP712 transaction, 
- `payloadSigner`: provides a way of  writing arbitrary logic for signing transactions, messages, typed data.

This enables complete flexibility and thus support for Account Abstraction. Also, the smart account no longer has the concept of `keys`, but rather an abstract way in the `secret` to support other way of signing instead of just ECDSA. That's why `ethers.SigningKey` and `string | string[]` types are replaced with `any`. This secret is mandatory and is used for signing transactions and other playloads. By default, `SmartAccount` uses  a single ECDSA private key as a `secret`  and works in the identical way as `Wallet`. That means it's  compatible with `ethers.ContractFactory` and `ethers.Contract` classes. 

Given that the `SmartAccount` is flexible enough to support any methods, in order to provide a better user experience, implementations for commonly used cases have been created in the utils module. Implementations are provided for single ECDSA key for signing, and signing with multiple  ECDSA keys (multisig). See the `smart-account-utils.ts` for function implementations.  

Examples on how to create `SmartAccount`:

```ts
const account = new SmartAccount({ address, secret }, provider); // single ECDSA used for signing

const account = return new SmartAccount( // uses multiple ECDSA for singing payloads
       return new SmartAccount(
      {
        address,
        secret,
        payloadSigner: signPayloadWithMultipleECDSA,
        transactionBuilder: populateTransactionMultisigECSDA,
      },
      provider
);
``` 

Additionally, two factory classes have been created to simplify the creation of 'SmartAccount' and prevent misconfiguration by using the wrong function for signing (`ECDSASmartAccount`, `MultisigECDSASmartAccount`). Users can also create their own factories to bind 'SmartAccount' with appropriate custom functions.

`SmartAccount` also simplifies the way of interacting with account and contracts:

Example of interacting with mulstisig account using `Wallet`:

```ts
    let aaTx = await storage.set.populateTransaction(500);
    aaTx.from = await owner2.getAddress();
    const gasLimit = await provider.estimateGas(aaTx);
    const gasPrice = await provider.getGasPrice();

    aaTx = {
        ...aaTx,
        from: multisigAddress,
        gasLimit,
        gasPrice,
        chainId: (await provider.getNetwork()).chainId,
        nonce: await provider.getTransactionCount(multisigAddress),
        type: 113,
        customData: {
            gasPerPubdata: zk.utils.DEFAULT_GAS_PER_PUBDATA_LIMIT,
        } as zk.types.Eip712Meta,
        value: ethers.toBigInt(0),
    };
    const signedTxHash = zk.EIP712Signer.getSignedDigest(aaTx);

    const signature = ethers.concat([
        // Note, that `signMessage` wouldn't work here, since we don't want
        // the signed hash to be prefixed with `\x19Ethereum Signed Message:\n`
        ethers.Signature.from(owner1.signingKey.sign(signedTxHash)).serialized,
        ethers.Signature.from(owner2.signingKey.sign(signedTxHash)).serialized,
    ]);

    aaTx.customData = {
        ...aaTx.customData,
        customSignature: signature,
    };

    const serialized = zk.utils.serializeEip712(aaTx);

    const tx = await provider.broadcastTransaction(serialized);
    await tx.wait();
```
The same example using the SmartAccount:

```ts
const storageSetTx = await storage.set.populateTransaction(500);

const multisigAccount = MultisigECDSASmartAccount.create(
      multisigAddress,
      [owner1.signingKey, owner2.signingKey],
      provider,
);

const tx = await multisigAccount.sendTransaction({ ...storageSetTx });
await tx.wait();
``` 

`SmartAccount` also works with `ethers.ContractFactory`, `ethers.Contract` and provides paymaster support:

```ts
const account = ECDSASmartAccount.create(ADDRESS, PRIVATE_KEY , provider);

const storageFactory = new ContractFactory(storageAbi, storageBytecode, account);
const storage = (await storageFactory.deploy()) as Contract;

const paymasterSetTx = await storage.set(700, {
      customData: {
          paymasterParams: utils.getPaymasterParams(PAYMASTER, {
              type: "ApprovalBased",
              token: TOKEN,
              minimalAllowance: 1,
              innerInput: new Uint8Array(),
          }),
        },
     });
await paymasterSetTx.wait();
```

### CODE REVIEW

**Comment by dutterbutter** [comment]
@danijelTxFusion can you resolve conflicts.

**Comment by dutterbutter** [comment]
@danijelTxFusion can we remove the "customData" object entirely and have the params inline?

**Comment by danijelTxFusion** [comment]
> @danijelTxFusion can we remove the "customData" object entirely and have the params inline?

Unfortunately, no. To support inline parameters, we need to duplicate most of the `Contract` class and helper functions to customize the logic because those functions are private. It would be as if we created a mini-fork in our SDK, and updates in ethers related to `Contract` wouldn't be taken into account, increasing maintenance effort. However, the `overrides.customData` is a well-known concept in `ethers`, and the community has adopted this approach. I don't think this will be a problem.

**Comment by dutterbutter** [comment]
> > @danijelTxFusion can we remove the "customData" object entirely and have the params inline?
> 
> @dutterbutter Unfortunately, no. To support inline parameters, we need to duplicate most of the `Contract` class and helper functions to customize the logic because those functions are private. It would be as if we created a mini-fork in our SDK, and updates in `ethers` related to `Contract` wouldn't be taken into account, increasing maintenance effort. However, the `overrides.customData` is a well-known concept in `ethers`, and the community has adopted this approach. I don't think this will be a problem.

Makes sense, do need to go against the grain.

**Comment by dutterbutter** [comment]
This was a much a better experience then previously!! I want to try a couple different use cases before approving but so far a much better experience. Do link the docs PR here once that is up! 🙏 Thanks @danijelTxFusion

**Comment by dutterbutter** [comment]
@danijelTxFusion quick question, if we merge this in will old approaches like the one described in the [existing multi-sig tutorial ](https://docs.zksync.io/build/tutorials/smart-contract-development/account-abstraction/custom-aa-tutorial.html#start-a-transaction-from-the-account) still work as intended? 

Understandably, we will re-write this tutorial to reflect the new easy approach but that would likely be in the future.

File: src/smart-account-utils.ts [hunk 1]
```diff
@@ -0,0 +1,264 @@
+import {ethers, SigningKey} from 'ethers';
+import {EIP712_TX_TYPE, DEFAULT_GAS_PER_PUBDATA_LIMIT} from './utils';
+import {TransactionLike, TransactionBuilder, PayloadSigner} from './types';
+
+/**
+ * Sings the `payload` using an ECDSA private key.
```
Comment (dutterbutter) [inline src/smart-account-utils.ts:6]:
```suggestion
 * Signs the `payload` using an ECDSA private key.
```

File: src/smart-account-utils.ts [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+ */
+export const signPayloadWithECDSA: PayloadSigner = async (
+  payload,
+  secret: string | SigningKey
+) => {
+  return new ethers.Wallet(secret).signingKey.sign(payload).serialized;
+};
+
+/**
+ * Sings the `payload` using multiple ECDSA private keys.
```
Comment (dutterbutter) [inline src/smart-account-utils.ts:66]:
```suggestion
 * Signs the `payload` using multiple ECDSA private keys.
```

**Review by dutterbutter — APPROVED** [review]
@danijelTxFusion it would be good if someone else from txFusion also reviews the PR. But LGTM.

**Review by mpopovac-txfusion — APPROVED** [review]

File: src/smart-account-utils.ts (hunk 1 above)
Comment (danijelTxFusion) [reply src/smart-account-utils.ts]: Resolved.

**Pull request merged**